test/Microsoft.ML.Tests/CollectionDataSourceTests.cs

```diff
@@ -205,5 +207,314 @@ public class IrisPrediction
+                    fBool =new bool[2]{true, false }, fByte = new byte[3]{ 0,124,255}, fDouble=new double[3]{ -1,0, 1}, fLong = new long[]{ 0,1,2} ,
+                    fsByte = new sbyte[3]{ -127,127,0}, fShort = new short[3]{ 0, 1225, 32767 }, fuInt =new uint[2]{ 0, uint.MaxValue},
+                    fuLong = new ulong[2]{ ulong.MaxValue, 0}, fuShort = new ushort[2]{ 0, ushort.MaxValue}
+                },
+                new ClassWithArrays(){ fInt = new int[3]{ -2,1,0}, fFloat = new float[3]{ 0.99f, 0f, -0.99f}, fString =new string[2]{"",null} },
+                new ClassWithArrays()
+            };
+
+            var nullableData = new List<ClassWithNullableArrays>()
+            {
+                new ClassWithNullableArrays(){ fInt = new int?[3]{ null,-1,1}, fFloat = new float?[3]{ -0.99f, null, 0.99f}, fString =new string[2]{ null, ""},
+                    fBool =new bool?[3]{true,null, false }, fByte = new byte?[4]{ 0,125,null,255}, fDouble=new double?[3]{ -1,null, 1}, fLong = new long?[]{null,-1,1} ,
+                    fsByte = new sbyte?[3]{ -127,127,null}, fShort = new short?[3]{ 0, null, 32767 }, fuInt =new uint?[4]{null,42 ,0, uint.MaxValue},
+                    fuLong = new ulong?[3]{ ulong.MaxValue, null, 0}, fuShort = new ushort?[3]{ 0,null, ushort.MaxValue}
+                },
+                new ClassWithNullableArrays(){ fInt = new int?[3]{ -2,1,0}, fFloat = new float?[3]{ 0.99f, 0f, -0.99f}, fString =new string[2]{  "lola","hola"} },
+                new ClassWithNullableArrays()
+            };
+
+            using (var env = new TlcEnvironment())
+            {
+                var dataView = ComponentCreation.CreateDataView(env, data);
+                var enumeratorSimple = dataView.AsEnumerable<ClassWithArrays>(env, false).GetEnumerator();
+                var originalEnumerator = data.GetEnumerator();
+                while (enumeratorSimple.MoveNext() && originalEnumerator.MoveNext())
+                {
+                    Assert.True(CompareThrougReflection(enumeratorSimple.Current, originalEnumerator.Current));
+                }
+                Assert.True(!enumeratorSimple.MoveNext() && !originalEnumerator.MoveNext());
+
+                var nullableDataView = ComponentCreation.CreateDataView(env, nullableData);
+                var enumeratorNullable = nullableDataView.AsEnumerable<ClassWithNullableArrays>(env, false).GetEnumerator();
+                var originalNullalbleEnumerator = nullableData.GetEnumerator();
+                while (enumeratorNullable.MoveNext() && originalNullalbleEnumerator.MoveNext())
+                {
+                    Assert.True(CompareThrougReflection(enumeratorNullable.Current, originalNullalbleEnumerator.Current));
+                }
+                Assert.True(!enumeratorNullable.MoveNext() && !originalNullalbleEnumerator.MoveNext());
+            }
+        }
```

Comment:
>} [](start = 8, length = 1)

Good job on the new tests. I assume you ran test coverage of them, right?

Reply:
yep, all new changes are covered

---
In reply to: [203883396](https://github.com/dotnet/machinelearning/pull/555#discussion_r203883396) [](ancestors = 203883396)